File: internal/git/commit.go
package git

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// StageAllFiles stages all files in the repository
func StageAllFiles(path string) error {
	absPath, err := filepath.Abs(path)
	if err != nil {
		return fmt.Errorf("failed to get absolute path: %w", err)
	}

	workDir := absPath
	stat, err := os.Stat(absPath)
	if err == nil && !stat.IsDir() {
		workDir = filepath.Dir(absPath)
	}

	cmd := exec.Command("git", "add", ".")
	cmd.Dir = workDir

	if err := cmd.Run(); err != nil {
		return fmt.Errorf("failed to stage all files: %w", err)
	}

	return nil
}

// CreateCommit creates a commit with the given message
func CreateCommit(path, message string) error {
	absPath, err := filepath.Abs(path)
	if err != nil {
		return fmt.Errorf("failed to get absolute path: %w", err)
	}

	workDir := absPath
	stat, err := os.Stat(absPath)
	if err == nil && !stat.IsDir() {
		workDir = filepath.Dir(absPath)
	}

	cmd := exec.Command("git", "commit", "-m", message)
	cmd.Dir = workDir

	if err := cmd.Run(); err != nil {
		return fmt.Errorf("failed to create commit: %w", err)
	}

	return nil
}

// HasStagedChanges checks if there are any staged changes
func HasStagedChanges(path string) (bool, error) {
	absPath, err := filepath.Abs(path)
	if err != nil {
		return false, fmt.Errorf("failed to get absolute path: %w", err)
	}

	workDir := absPath
	stat, err := os.Stat(absPath)
	if err == nil && !stat.IsDir() {
		workDir = filepath.Dir(absPath)
	}

	cmd := exec.Command("git", "diff", "--staged", "--quiet")
	cmd.Dir = workDir

	err = cmd.Run()
	// Exit code 1 means there are staged changes
	if exitErr, ok := err.(*exec.ExitError); ok && exitErr.ExitCode() == 1 {
		return true, nil
	}
	// Exit code 0 means no staged changes
	if err == nil {
		return false, nil
	}

	return false, fmt.Errorf("failed to check staged changes: %w", err)
}

// GetStagedFiles returns the list of staged files
func GetStagedFiles(path string) ([]string, error) {
	absPath, err := filepath.Abs(path)
	if err != nil {
		return nil, fmt.Errorf("failed to get absolute path: %w", err)
	}

	workDir := absPath
	stat, err := os.Stat(absPath)
	if err == nil && !stat.IsDir() {
		workDir = filepath.Dir(absPath)
	}

	cmd := exec.Command("git", "diff", "--staged", "--name-only")
	cmd.Dir = workDir

	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("failed to get staged files: %w", err)
	}

	files := strings.Split(strings.TrimSpace(string(output)), "\n")
	if len(files) == 1 && files[0] == "" {
		return []string{}, nil
	}

	return files, nil
}

// PushBranch pushes the current branch to the remote repository
func PushBranch(path string) error {
	absPath, err := filepath.Abs(path)
	if err != nil {
		return fmt.Errorf("failed to get absolute path: %w", err)
	}

	workDir := absPath
	stat, err := os.Stat(absPath)
	if err == nil && !stat.IsDir() {
		workDir = filepath.Dir(absPath)
	}

	cmd := exec.Command("git", "push")
	cmd.Dir = workDir

	if err := cmd.Run(); err != nil {
		return fmt.Errorf("failed to push branch: %w", err)
	}

	return nil
}

// GetCurrentBranch returns the current branch name
func GetCurrentBranch(path string) (string, error) {
	absPath, err := filepath.Abs(path)
	if err != nil {
		return "", fmt.Errorf("failed to get absolute path: %w", err)
	}

	workDir := absPath
	stat, err := os.Stat(absPath)
	if err == nil && !stat.IsDir() {
		workDir = filepath.Dir(absPath)
	}

	cmd := exec.Command("git", "branch", "--show-current")
	cmd.Dir = workDir

	output, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("failed to get current branch: %w", err)
	}

	return strings.TrimSpace(string(output)), nil
}

File: critica_py/git/commit.py
"""Git commit operations."""

import subprocess
from pathlib import Path
from typing import List


def stage_all_files(path: str = ".") -> None:
    """Stage all files in the repository.

    Args:
        path: Repository path (default: current directory)
    """
    abs_path = Path(path).resolve()

    if abs_path.is_file():
        work_dir = abs_path.parent
    else:
        work_dir = abs_path

    try:
        subprocess.run(
            ["git", "add", "."], cwd=work_dir, capture_output=True, text=True, check=True
        )
    except subprocess.CalledProcessError as e:
        raise RuntimeError(f"Failed to stage all files: {e}")


def create_commit(path: str, message: str) -> None:
    """Create a commit with the given message.

    Args:
        path: Repository path
        message: Commit message
    """
    abs_path = Path(path).resolve()

    if abs_path.is_file():
        work_dir = abs_path.parent
    else:
        work_dir = abs_path

    try:
        subprocess.run(
            ["git", "commit", "-m", message],
            cwd=work_dir,
            capture_output=True,
            text=True,
            check=True,
        )
    except subprocess.CalledProcessError as e:
        raise RuntimeError(f"Failed to create commit: {e}")


def has_staged_changes(path: str = ".") -> bool:
    """Check if there are any staged changes.

    Args:
        path: Repository path (default: current directory)

    Returns:
        True if there are staged changes, False otherwise
    """
    abs_path = Path(path).resolve()

    if abs_path.is_file():
        work_dir = abs_path.parent
    else:
        work_dir = abs_path

    try:
        result = subprocess.run(
            ["git", "diff", "--staged", "--quiet"],
            cwd=work_dir,
            capture_output=True,
            text=True,
            check=False,
        )
        # Exit code 1 means there are staged changes
        # Exit code 0 means no staged changes
        return result.returncode == 1
    except subprocess.CalledProcessError as e:
        raise RuntimeError(f"Failed to check staged changes: {e}")


def get_staged_files(path: str = ".") -> List[str]:
    """Get list of staged files.

    Args:
        path: Repository path (default: current directory)

    Returns:
        List of staged file paths
    """
    abs_path = Path(path).resolve()

    if abs_path.is_file():
        work_dir = abs_path.parent
    else:
        work_dir = abs_path

    try:
        result = subprocess.run(
            ["git", "diff", "--staged", "--name-only"],
            cwd=work_dir,
            capture_output=True,
            text=True,
            check=True,
        )

        files = [f for f in result.stdout.strip().split("\n") if f]
        return files
    except subprocess.CalledProcessError as e:
        raise RuntimeError(f"Failed to get staged files: {e}")


def push_branch(path: str = ".") -> None:
    """Push the current branch to the remote repository.

    Args:
        path: Repository path (default: current directory)
    """
    abs_path = Path(path).resolve()

    if abs_path.is_file():
        work_dir = abs_path.parent
    else:
        work_dir = abs_path

    try:
        subprocess.run(["git", "push"], cwd=work_dir, capture_output=True, text=True, check=True)
    except subprocess.CalledProcessError as e:
        raise RuntimeError(f"Failed to push branch: {e}")


def get_recent_commits(path: str = ".", count: int = 5) -> List[dict]:
    """Get recent commit messages.

    Args:
        path: Repository path (default: current directory)
        count: Number of commits to retrieve (default: 5)

    Returns:
        List of dictionaries with commit information
    """
    abs_path = Path(path).resolve()

    if abs_path.is_file():
        work_dir = abs_path.parent
    else:
        work_dir = abs_path

    try:
        # Format: hash|author|date|message
        result = subprocess.run(
            ["git", "log", f"-{count}", "--pretty=format:%h|%an|%ar|%s"],
            cwd=work_dir,
            capture_output=True,
            text=True,
            check=True,
        )

        commits = []
        for line in result.stdout.strip().split("\n"):
            if not line:
                continue
            parts = line.split("|", 3)
            if len(parts) == 4:
                commits.append(
                    {"hash": parts[0], "author": parts[1], "date": parts[2], "message": parts[3]}
                )

        return commits
    except subprocess.CalledProcessError:
        return []
